### ISSUE TITLE
Is there a way for multimarkdown to NOT suppress the TODOs?

### ISSUE THREAD

**Issue by neochief** [vsch/idea-multimarkdown#309]
I put a lot of TODO marks into my markdown files. Then I use the default TODO explorer to navigate through them and fix issues one by one. After installing multimarkdown it seems that TODOs in .md files are not longer recognized. Here are the TODO fingerprints I used to use:

// TODO: some text
>TODO: some text

Is there a workaround to turn them on again? I use all defaults in settings.

**Comment by vsch** [comment]
The IDE provided TODO tool window displays TODO or other labels found in comments.

To recognize TODO tags in markdown you need to wrap them in HTML comments:

    <!-- TODO: some text !>

<img width="764" alt="screen shot 2016-09-27 at 10 01 45 am" src="https://cloud.githubusercontent.com/assets/10299086/18876915/7818dcd2-8499-11e6-8058-723a66d3d985.png">

**Comment by neochief** [comment]
Okay, thanks for the workaround. I guess IDE treats the whole file as a comment when dealing with unrecognized file type (such as it was before I installed multimarkdown).

**Comment by habemus-papadum** [comment]
In scaffolding out large markdown file(s), it would be nice to have a more concise notation for TODO's then embedding them into html comments e.g. allow `//TODO ...` vs  `<-- TODO ... -->`

**Comment by vsch** [comment]
@habemus-papadum, the IDE requires TODO to be part of a comment. Markdown does not have comments other than HTML.

Introducing `//` or other type of comment mechanism to markdown will make it non-standard and will result in every other markdown processor rendering it as plain text.

**Comment by habemus-papadum** [comment]
makes sense.  I will add a smart template to insert e.g. `<!-- TODO: -->` .  Would you consider teaching the markdown editor to "squash" the todo comments in the way the intellij editors do for generics in Java and templates in C++?
e.g. the editor would abbreviate `<!-- TODO: -->` as `TODO:`

**Comment by vsch** [comment]
@habemus-papadum, that is a good idea. I will add it to markdown folding. 

Should the folded placeholder include `TODO: with comment text` or just `TODO: ...`?

**Comment by habemus-papadum** [comment]
My style would be to use , as the text that follows is likely a Markdown heading or list item fragment, and the comment is a note to myself to this item needs extra work.  But I'm sure others might find use for the alternative.   Maybe there is a clever solution that supports both usages with one implementation.  But thanks very much for considering the suggestion.  Cheers, nehal

**Comment by vsch** [comment]
@habemus-papadum, the difference between the two options is:

1. results in shorter text but you need to hover over the folded text to see details
2. results in longer text but the comment is visible so no added action is needed to see the comment.

There is no clever solution to have both because the option needs to be selected by the user. Either a single option controls the folding and the type of folding is hard-coded or two options need to be provided to allow selecting either none for no folding or one option for short text or long text, both options selected will result in short text.

**Comment by habemus-papadum** [comment]
I would start with 1., but allow the case where the wrapped text after TODO: is empty (i.e. my use case).   Later on  maybe add a parameter that let's users select a max size, which causes long text to be elided after a certain number of characters.

**Comment by vsch** [comment]
@habemus-papadum, the empty TODO: text will be handled automatically since the text will be everything between comment tags which has the regex form `\w+:.*`.

At least for the first implementation I will not use the IDE TODO regex settings to match markdown comments to actual TODO definitions. The IDE allows any regex to be used for the matches but that will be ignored for now.

**Comment by habemus-papadum** [comment]
That sounds fine -- I noticed IDE TODO regex's and  wondered if the plugin API allowed for querying those values -- I don't really think that is much of a usage issue

**Comment by vsch** [comment]
EAP update with a fix for the issue is available.

To enable EAP update channel:

1. Select "Early Access Program" in `Check updates for:`
2. Then make the IDE check for updates.

![Screen Shot Settings Update Channel](https://github.com/vsch/idea-multimarkdown/raw/master/assets/images/faq/ScreenShot_Settings_UpdateChannel.png)

**Comment by vsch** [comment]
@habemus-papadum, need to enable `Markdown Comments` in Editor > General > Code Folding

**Comment by habemus-papadum** [comment]
- updated to latest EAP
- checked `Markdown Comments` in Editor > General > Code Folding

I see:
![screen shot 2018-06-29 at 12 59 23 pm](https://user-images.githubusercontent.com/1161604/42104947-a04c7afa-7b9c-11e8-9c45-7993c92a3409.png)

no folding, and maybe an issue with tooltips?  Let me know what to provide to help debug

**Comment by habemus-papadum** [comment]
There are a stack trace:

```
java.lang.Throwable: Invalid range specified: (12, 11); 
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:123)
	at com.intellij.openapi.util.TextRange.assertProperRange(TextRange.java:229)
	at com.intellij.openapi.util.TextRange.assertProperRange(TextRange.java:224)
	at com.intellij.openapi.util.TextRange.assertProperRange(TextRange.java:220)
	at com.intellij.openapi.util.TextRange.<init>(TextRange.java:43)
	at com.intellij.openapi.util.TextRange.<init>(TextRange.java:32)
	at com.vladsch.idea.multimarkdown.language.folding.MultiMarkdownFoldingBuilder.buildFoldRegions(MultiMarkdownFoldingBuilder.kt:350)
	at com.intellij.lang.folding.LanguageFolding.buildFoldingDescriptors(LanguageFolding.java:71)
	at com.intellij.codeInsight.folding.impl.FoldingUpdate.a(FoldingUpdate.java:239)
	at com.intellij.codeInsight.folding.impl.FoldingUpdate.getFoldingsFor(FoldingUpdate.java:225)
	at com.intellij.codeInsight.folding.impl.FoldingUpdate.a(FoldingUpdate.java:97)
	at com.intellij.codeInsight.folding.impl.FoldingUpdate.a(FoldingUpdate.java:86)
	at com.intellij.psi.impl.PsiParameterizedCachedValue.doCompute(PsiParameterizedCachedValue.java:45)
	at com.intellij.util.CachedValueBase.getValueWithLock(CachedValueBase.java:222)
	at com.intellij.psi.impl.PsiCachedValue.getValueWithLock(PsiCachedValue.java:60)
	at com.intellij.psi.impl.PsiParameterizedCachedValue.getValue(PsiParameterizedCachedValue.java:35)
	at com.intellij.psi.util.CachedValuesManager.getParameterizedCachedValue(CachedValuesManager.java:96)
	at com.intellij.codeInsight.folding.impl.FoldingUpdate.updateFoldRegions(FoldingUpdate.java:82)
	at com.intellij.codeInsight.folding.impl.CodeFoldingManagerImpl.a(CodeFoldingManagerImpl.java:298)
	at com.intellij.codeInsight.folding.impl.CodeFoldingManagerImpl.updateFoldRegions(CodeFoldingManagerImpl.java:257)
	at com.intellij.codeInsight.folding.impl.CodeFoldingManagerImpl.updateFoldRegions(CodeFoldingManagerImpl.java:251)
	at com.intellij.codeInsight.folding.impl.actions.ExpandRegionAction.expandRegionAtOffset(ExpandRegionAction.java:48)
	at com.intellij.codeInsight.folding.impl.actions.ExpandRegionAction.a(ExpandRegionAction.java:43)
	at com.intellij.codeInsight.folding.impl.actions.ExpandRegionAction.access$000(ExpandRegionAction.java:30)
	at com.intellij.codeInsight.folding.impl.actions.ExpandRegionAction$1.doExecute(ExpandRegionAction.java:35)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.lambda$execute$4(EditorActionHandler.java:207)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.doIfEnabled(EditorActionHandler.java:110)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.execute(EditorActionHandler.java:206)
	at com.intellij.openapi.editor.actionSystem.EditorAction.lambda$actionPerformed$0(EditorAction.java:96)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.a(CoreCommandProcessor.java:139)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:97)
	at com.intellij.openapi.editor.actionSystem.EditorAction.actionPerformed(EditorAction.java:105)
	at com.intellij.openapi.editor.actionSystem.EditorAction.actionPerformed(EditorAction.java:80)
	at com.intellij.openapi.actionSystem.ex.ActionUtil$1.run(ActionUtil.java:220)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performActionDumbAware(ActionUtil.java:237)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher$1.performAction(IdeKeyEventDispatcher.java:620)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.a(IdeKeyEventDispatcher.java:669)
	at com.intellij.openapi.application.TransactionGuardImpl.performUserActivity(TransactionGuardImpl.java:195)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processAction(IdeKeyEventDispatcher.java:668)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.b(IdeKeyEventDispatcher.java:520)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.f(IdeKeyEventDispatcher.java:475)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.dispatchKeyEvent(IdeKeyEventDispatcher.java:212)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:697)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:382)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
```

**Comment by habemus-papadum** [comment]
Exception is fixed.

This is what I see
![screen shot 2018-06-29 at 1 33 39 pm](https://user-images.githubusercontent.com/1161604/42106230-4bbdd754-7ba1-11e8-9d02-6cfb3f7d4730.png)

Small points -- perhaps for another day...  

-  Did you intend for the trailing `-->` to be visible.  It actually is not ugly, but probably should be elided.
-  Is it possible to keep the TODO yellow, italics when folded?
- I have to manually fold (`Ctrl - -`) the TODO for it to disappear - ideally the editor's parser would autofold the TODO after creation (similar to what Clion does after creating a C++ template declaration)

Thanks for your assistance!

**Comment by vsch** [comment]
@habemus-papadum,

1. The trailing --> seems to be an issue with ligatures. I have to look into the API for ligatures because the folding handles text without ligatures only. I don't have ligatures enabled and the trailing --> does not show when folded.
2. The folding text does not seem to have highlighting options other than those for folded text
3. The IDE remembers which folding regions were expanded. Selecting `Markdown Comments` in Editor > General > Code Folding will make the comments fold by default.

**Comment by habemus-papadum** [comment]
@vsch 
> 1. The trailing -->

If I turn off ligatures I still get the same behavior, but see below...

>  2. highlighting options

Understood

> 3. IDE remembers which folding regions

This isn't happening.  If I create several todos (by explicit typing, copy and paste, or live template) they all are created unfolded even though `Markdown Comments` is selected.  I can manually fold, but then I get a trailing `-->` whether ligatures are on or not.  If I manually toggle off  `Markdown Comments`, and then toggle it on, the editor folds all the todos, with no  trailing `-->` (In this instance I tested only ligatures off.  
This makes it possible for me to create all sorts of variations as shown below:
![screen shot 2018-06-29 at 2 03 27 pm](https://user-images.githubusercontent.com/1161604/42107412-3c023a0e-7ba5-11e8-976b-da79f90e8117.png)

It's possible my webstorm

```
WebStorm 2018.2 EAP
Build #WS-182.3458.21, built on June 28, 2018
WebStorm EAP User
Expiration date: July 28, 2018
JRE: 1.8.0_152-release-1226-b7 x86_64
JVM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
macOS 10.13.3
 ```
Has managed to get itself into a funky state

**Comment by vsch** [comment]
@habemus-papadum, I tried it on the same version WebStorm latest 2018.EAP and cannot get the trailing --> to show when folded:

<img width="261" alt="image" src="https://user-images.githubusercontent.com/10299086/42113573-aeb5c764-7bb9-11e8-8042-7aa503febd18.png">

If you have a markdown sample that exhibits this behaviour it would help if I tried it on my system. It could be something else in the file causing this.

**Comment by habemus-papadum** [comment]
@vsch 
I get the behavior even if the todo is the only line in the file:

![screen shot 2018-06-29 at 5 13 14 pm](https://user-images.githubusercontent.com/1161604/42115140-c4b4efd0-7bbf-11e8-918a-fbddaacb8ed8.png)

If the IDE looses focus (blurs), then the `-->` suddenly goes away. 

![screen shot 2018-06-29 at 5 11 19 pm](https://user-images.githubusercontent.com/1161604/42115044-74878572-7bbf-11e8-9289-ca99295d33a6.png)

Kind of odd behavior -- I am going to see if this happens in Clion

**Comment by habemus-papadum** [comment]
Hi  --  I tried in Clion and RubyMine and got strange results that are difficult to summarize.  

Ultimately I was able to finds ways to get into a broken state , as well as avoid getting into broken states if I tried hard.  My best guess is that I think ligatures are problematic, and once the editor hits issues with ligatures it gets into a weird state that does not resolve until restart.  

I think for now I can live without folding of TODO's -- in few days I will revisit and see if I can create a minimal reproducible test case and will revert back with more information.  

Thanks!
